dds/DCPS/XTypes/DynamicDataImpl.cpp

```diff
@@ -1065,53 +1467,53 @@ bool DynamicDataImpl::set_value_to_union(DDS::MemberId id, const MemberType& val
         return false;
       }
 
-      CORBA::Long disc_value;
-      if (!cast_to_discriminator_value(disc_value, value)) {
-        return false;
-      }
-
-      if (discriminator_selects_no_member(type_, disc_value)) {
-        if (reset_union() != DDS::RETCODE_OK) {
-          return false;
-        }
-      } else if (!validate_discriminator(disc_value, selected_md)) {
+      if (!validate_discriminator(disc_value, selected_md)) {
         if (DCPS::log_level >= DCPS::LogLevel::Notice) {
           ACE_ERROR((LM_NOTICE, "(%P|%t) NOTICE: DynamicDataImpl::set_value_to_union:"
-                     " Discriminator value %d does not select existing selected member (ID %d)\n",
+                     " Discriminator value %d does not select the activated member (ID %u)\n",
                      disc_value, selected_id));
         }
         return false;
       }
-    }
-
-  } else { // Writing a selected member
-    if (reset_union() != DDS::RETCODE_OK) {
-      return false;
-    }
-    DDS::DynamicTypeMember_var member;
-    if (type_->get_member(member, id) != DDS::RETCODE_OK) {
-      return false;
-    }
-    DDS::MemberDescriptor_var md;
-    if (member->get_descriptor(md) != DDS::RETCODE_OK) {
-      return false;
-    }
-    member_type = get_base_type(md->type());
-
-    const TypeKind member_tk = member_type->get_kind();
-    if (member_tk != MemberTypeKind && member_tk != enum_or_bitmask) {
-      return false;
-    }
-
-    if (!insert_valid_discriminator(md)) {
+      return insert_single(id, value);
+    } else {
+      // In case the union has implicit default member and the input discriminator value
+      // selects that implicit default member, store the discriminator value. The semantics
+      // of this is similar to the _default() method of the IDL-to-C++ mapping for union.
+      if (discriminator_selects_no_member(type_, disc_value)) {
+        return insert_single(id, value);
+      }
+      if (DCPS::log_level >= DCPS::LogLevel::Notice) {
```

Comment:
Upgrade this to an error.

Reply:
The table in the guidelines has `LogLevel::Error` for unrecoverable error but this one looks recoverable to me. It will also return `DDS::RETCODE_ERROR` to the user, so maybe it's also reportable even though `DDS::RETCODE_ERROR` is not specific enough. But I'm not sure if it's really considered reportable in this case. If it's not, perhaps move it to `LogLevel::Warning`?

Reply:
If it's a `ReturnCode_t` the user gets (or ultimately becomes one), it should be almost certainly be logged using notice. There could be an exception for code that's shared with something that's not returned to the user, but in that case it's probably more proper to have a separate warning/error at non-user caller.

Reply:
Sounds like using notice is what we want here since the user will receive `ReturnCode_t` in this case.

Reply:
You are correct on this.